src/kokkos/plugin-SiPixelClusterizer/kokkos/gpuClusterChargeCut.h

```diff
@@ -4,126 +4,149 @@
-      for (auto i = threadIdx.x; i < nclus; i += blockDim.x) {
-        charge[i] = 0;
-      }
-      __syncthreads();
-
-      for (auto i = first; i < numElements; i += blockDim.x) {
-        if (id[i] == InvId)
-          continue;  // not valid
-        if (id[i] != thisModuleId)
-          break;  // end of module
-        atomicAdd(&charge[clusterId[i]], adc[i]);
-      }
-      __syncthreads();
-
-      auto chargeCut = thisModuleId < 96 ? 2000 : 4000;  // move in constants (calib?)
-      for (auto i = threadIdx.x; i < nclus; i += blockDim.x) {
-        newclusId[i] = ok[i] = charge[i] > chargeCut ? 1 : 0;
-      }
-
-      __syncthreads();
-
-      // renumber
-      __shared__ uint16_t ws[32];
-      blockPrefixScan(newclusId, nclus, ws);
-
-      assert(nclus >= newclusId[nclus - 1]);
-
-      if (nclus == newclusId[nclus - 1])
-        return;
-
-      nClustersInModule[thisModuleId] = newclusId[nclus - 1];
-      __syncthreads();
-
-      // mark bad cluster again
-      for (auto i = threadIdx.x; i < nclus; i += blockDim.x) {
-        if (0 == ok[i])
-          newclusId[i] = InvId + 1;
-      }
-      __syncthreads();
-
-      // reassign id
-      for (auto i = first; i < numElements; i += blockDim.x) {
-        if (id[i] == InvId)
-          continue;  // not valid
-        if (id[i] != thisModuleId)
-          break;  // end of module
-        clusterId[i] = newclusId[clusterId[i]] - 1;
-        if (clusterId[i] == InvId)
-          id[i] = InvId;
-      }
+          charge_view_type charge(teamMember.team_scratch(shared_view_level),::gpuClustering::MaxNumClustersPerModules);
+          ok_view_type ok(teamMember.team_scratch(shared_view_level),::gpuClustering::MaxNumClustersPerModules);
+          newclusid_view_type newclusId(teamMember.team_scratch(shared_view_level),::gpuClustering::MaxNumClustersPerModules);
+
+          assert(nclus <= ::gpuClustering::MaxNumClustersPerModules);
+          for (uint32_t i = teamMember.team_rank(); i < ::gpuClustering::MaxNumClustersPerModules; i += teamMember.team_size()) {
+            charge(i) = 0;
+            ok(i) = 0;
+            newclusId(i) = 0;
+          }
+          teamMember.team_barrier();
+
+          for (int i = first; i < numElements; i += teamMember.team_size()) {
+            if (id(i) == ::gpuClustering::InvId)
+              continue;  // not valid
+            if (id(i) != thisModuleId)
+              break;  // end of module
+            Kokkos::atomic_fetch_add(&charge(clusterId(i)), adc(i));
+          }
+          teamMember.team_barrier();
+
+          auto chargeCut = thisModuleId < 96 ? 2000 : 4000;  // move in constants (calib?)
+          for (uint32_t i = teamMember.team_rank(); i < nclus; i += teamMember.team_size()) {
+            newclusId(i) = ok(i) = charge(i) > chargeCut ? 1 : 0;
+          }
+          teamMember.team_barrier();
+
+          // renumber
+          Kokkos::single(Kokkos::PerTeam(teamMember),
```

Comment:
The challenge there is that this`blockPrefixScan` happens inside a parallel kernel. Kokkos does not support `parallel_scan` blocks inside `TeamPolicy` blocks. (Despite having an example of such code on their Wiki!)

Reply:
Right. But I suppose in principle one could split the kernel into three: everything before the scan, the scan, and everything after the scan (that we have done elsewhere), right? (I'm still not suggesting to do that, just wondering if it could be done)

Reply:
You can't in this case (like I did before), because of the shared memory objects that are used before and after. I guess I could again make global views that are arrays indexed by block number, but how much obfuscation do we want to introduce?

Reply:
Right, shared memory as well.

> how much obfuscation do we want to introduce?

Exactly, that's why I wasn't suggesting it. I just wanted to understand the implications **if** that would be done.